Always use -fasynchronous-unwind-table.

justinmichaud opened the pull request:
#### 7db7c2ae79a3a180ad03e10b71aed511eae4224b
<pre>
Always use -fasynchronous-unwind-table.
<a href="https://bugs.webkit.org/show_bug.cgi?id=279300">https://bugs.webkit.org/show_bug.cgi?id=279300</a>

Reviewed by Michael Catanzaro.

This option is needed on ARMv7 for libbacktrace to work at all. The
downstream port of WPE already has a similar patch, but it should be
upstreamed too.

* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/283378@main">https://commits.webkit.org/283378@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4c06a874c3399bc55415979a9ea3c97a61e911c2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66084 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45457 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18703 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70116 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16694 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53256 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16975 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53047 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11622 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69151 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41938 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57210 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33679 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38609 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15570 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/59198 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60503 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14934 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71818 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/65328 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10039 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14350 "Found 1 new test failure: http/wpt/mediastream/transfer-mediastreamtrack-to-worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60354 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10071 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57272 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60645 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14582 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8290 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1931 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/87095 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41265 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15325 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42341 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43524 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42085 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -416,6 +416,8 @@ if (USE_LIBBACKTRACE)
         message(FATAL_ERROR "libbacktrace is required for USE_LIBBACKTRACE")
     endif ()
+    set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -funwind-tables")
+    set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG}${CMAKE_COMPILER_SIZE_OPT_FLAGS} -funwind-tables")
```
OK, a few observations:

- OptionsGTK.cmake would need the same change
- You really don't want to add `${CMAKE_COMPILER_SIZE_OPT_FLAGS}` here. That's surely going to override whatever the desired optimization level is with -OS
- All major distros already build with -fasynchronous-unwind-tables. Do you know how adding -funwind-tables will interact with -fasynchronous-unwind-tables? I don't know what will happen, but I fear -funwind-tables may sabotage -fasynchronous-unwind-tables because you are appending it after the distro-provided flags.

Here are my suggestions:

 - Add the flag in WebKitCompilerFlags.cmake instead of here. It's OK to do so even when not building with libbacktrace.
 - Don't mess with size optimization flags, since that's unrelated
 - Use -fasynchronous-unwind-tables since that's what all distros do; presumably libbacktrace is compatible with that? Also, prepend this flag rather than appending it, since there's no reason to prevent overriding it.

q66 replied:
i would probably add this where `-fno-exceptions` is added, as this is only needed at all due to that

mcatanzaro commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -155,6 +155,7 @@ if (COMPILER_IS_GCC_OR_CLANG)
         WEBKIT_APPEND_GLOBAL_COMPILER_FLAGS(-fno-exceptions)
         WEBKIT_APPEND_GLOBAL_CXX_FLAGS(-fno-rtti)
         WEBKIT_APPEND_GLOBAL_CXX_FLAGS(-fcoroutines)
+        WEBKIT_APPEND_GLOBAL_CXX_FLAGS(-fasynchronous-unwind-tables)
```
```suggestion
        WEBKIT_PREPEND_GLOBAL_CXX_FLAGS(-fasynchronous-unwind-tables)
```

APPEND is right for the other flags here because we don't want those to be overridden. But it's harmless to override this one.

mcatanzaro reviewed (approved):
Please also update the commit message before landing, since you're now using the flag always (which is fine) rather than only when LIBBACKTRACE is enabled.

justinmichaud committed:
Always use -fasynchronous-unwind-table.
https://bugs.webkit.org/show_bug.cgi?id=279300

Reviewed by Michael Catanzaro.

This option is needed on ARMv7 for libbacktrace to work at all. The
downstream port of WPE already has a similar patch, but it should be
upstreamed too.

* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/283378@main
